MKL batched changes - API changes

ndellingwood opened the pull request:
Changes made to impl files; perf_test changes still needed.
Partially addresses issue #216

ndellingwood commented:
@kyungjoo-kim

kyungjoo-kim commented:
thanks.

kyungjoo-kim commented:
@ndellingwood Did you push your modification ? Or I should overwrite them ?

ndellingwood commented:
I only pushed the changes to the src files, not the perf_tests. Looks like the calls `LAPACKE_dgetrf_compact` in `KokkosBatched_LU_Serial_Impl.hpp` will need to be fixed also. Would you like me to push the perf_test changes also, or will those be clutter/junk?

kyungjoo-kim commented:
This is done. I will check spotcheck on blake and paste its result.

kyungjoo-kim commented:
I tested this on blake (SKX) and this is not related to my code. Am I the only one who see this failure ? I am usually unlucky person in spotcheck. So, it could be just me. 

```
[----------] 326 tests from openmp (223622 ms total)

[----------] Global test environment tear-down
[==========] 326 tests from 1 test case ran. (223622 ms total)
[  PASSED  ] 324 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] openmp.sparse_replaceSumIntoLonger_double_int64_t_int_TestExecSpace
[  FAILED  ] openmp.sparse_replaceSumIntoLonger_double_int64_t_size_t_TestExecSpace
```

This is configure output. 
```
Loading module intel/compilers/18.1.163
Loading module openblas
RUNNING: /home/kyukim/Work/lib/kokkoskernels/master/scripts/generate_makefile.bash \
--with-devices=OpenMP --arch=SKX --compiler=/home/projects/x86-64/intel/compilers/2018/compilers_and_libraries_2018.1.163/linux/bin/intel64/icpc \
--cxxflags="-O3 -Werror -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized " --ldflags="-O3 " \
--kokkos-path=/home/kyukim/Work/lib/kokkos/master --kokkoskernels-path=/home/kyukim/Work/lib/kokkoskernels/master --with-tpls=
Generating Makefiles with options  \
CXX=/home/projects/x86-64/intel/compilers/2018/compilers_and_libraries_2018.1.163/linux/bin/intel64/icpc \
KOKKOS_DEVICES=OpenMP \
KOKKOS_ARCH=SKX \
CXXFLAGS="-O3 -Werror -Wall -Wshadow -pedantic -Wsign-compare -Wtype-limits -Wuninitialized " \
LDFLAGS="-O3 " \
GTEST_PATH=/home/kyukim/Work/lib/kokkos/master/tpls/gtest \
KOKKOSKERNELS_OPTIONS=eti-only \
KOKKOS_PATH=/home/kyukim/Work/lib/kokkos/master \
KOKKOSKERNELS_PATH=/home/kyukim/Work/lib/kokkoskernels/master
```

srajama1 commented:
@ndellingwood : It is easier to explain this in person. Will do tomorrow.

ndellingwood commented:
I added updates to replace `mkl_?csrmultcsr` calls with `mkl_sparse_sp2m`. Removed the `float` versions during the numeric phase for now, but if that scalar type should be supported I can either write a struct templated on type with a static member function to wrap the call appropriate typed function, or just reinterpret_cast the pointer to force it to the proper type for compilation (though the wrong code path won't execute so there won't be problems). The reason for doing this is the existing if - elseif check must compile for both situations (is type float? is type double?) and thus fails with one or the other due to the pointer type mismatching. 

I can install kokkos-kernels as a lib without errors now on Bowman with intel/18.2.199.

ndellingwood commented:
@kyungjoo-kim or @srajama1 can you take a look at the changes I just pushed?

mhoemmen commented:
@ndellingwood It would be nice for the `Scalar = float` path to compile, even if it doesn't run correctly.

ndellingwood commented:
@mhoemmen will do, I'll just add some casting based on type, PR should be updated shortly.

ndellingwood commented:
Errors trying to build the perf_tests after the library built (On Bowman with intel/18.2.199 with OpenMP,Serial enabled):
```
icpc  KokkosSparse_pcg.o -L/home/ndellin/kokkos-kernels/testing/MKL-bowman/install/lib -L./ -lkokkos_kernels -L/mkl/lib/intel64 -lmkl_intel_lp64 -lmkl_intel_thread -lmkl_core -liomp5 -lpthread -lm -ldl -L/mkl/../compiler/lib/intel64 -lkokkos -ldl -lpthread -L/home/ndellin/kokkos-kernels/testing/MKL-bowman/kokkos/install/lib -xMIC-AVX512 -fopenmp  -o KokkosSparse_pcg.exe
ld: cannot find -lmkl_intel_lp64
ld: cannot find -lmkl_intel_thread
ld: cannot find -lmkl_core
make[1]: *** [KokkosSparse_pcg.exe] Error 1
```
Maybe the `Makefile.kokkos-kernels` needs an update for which libraries to link?

kyungjoo-kim commented:
This does not sound like a correct path ``-L/mkl/lib/intel64``
It should be something like 
```
[kyukim @bowman-lsm1] ~ > echo $MKLROOT
/home/projects/x86-64-knl/intel/compilers/2018/compilers_and_libraries_2018.2.199/linux/mkl
```

ndellingwood commented:
Ah, I included the path to the includes in the cxxflags when I generated my Makefile but never the library path.

ndellingwood commented:
Looked at Makefile.kokkos-kernels and the generate_makefile script, `MKL_PATH` is set by default using `SEMS_INTEL_ROOT` which doesn't exist on Bowman, so instead set the path on command line `export MKL_PATH=${MKLROOT}` for Bowman.

ndellingwood commented:
Pushed changes to allow for float and double types. Library builds, @kyungjoo-kim helped me get my Makefile generated properly to build the perf_tests and they all build now without compilation errors.

mhoemmen commented:
@ndellingwood Thanks for fixing `float`! :D

srajama1 commented:
Two things. Can we do a spot check MKL 17, so old code works ? Also a spot check on Haswell or Skylake where MKL 18 is available.

srajama1 commented on perf_test/batched/KokkosBatched_Test_Gemm_Host.hpp:
```diff
@@ -276,95 +276,74 @@ namespace KokkosBatched {
-            MKL_INT ldb[1] = { b.stride_1() };
-            MKL_INT ldc[1] = { c.stride_1() };
+            double done(1.0);
+            std::complex<double> zone(1.0);
```
Hopefully this perf_test path won't be compiled for CUDA ?

ndellingwood replied:
@srajama1 I looked a bit at this part of the test, it is all host code so only the host compiler should be compiling this, and zone is passed to the complex version of some mkl routine so using a `Kokkos::complex` would break the mkl interface. In addition, intel doesn't play with cuda, so we shouldn't run into a situation where this case could cause problems, good catch to check that spot though :)

srajama1 commented on perf_test/batched/KokkosBatched_Test_LU_Host.hpp:
```diff
@@ -17,9 +17,10 @@
-
     namespace PerfTest {
 
 #undef FLOP_MUL
```
Can we namespace this ? I know this is not part of this PR, still

ndellingwood replied:
@srajama1 The PerfTest namespace is still there, only the newline was removed. I added comments to mark end of each namespace as well, hopefully help the readability.

ndellingwood commented:
@srajama1 I updated the PR with a couple commit, main change was to the INTEL_MKL macro comparison for versioning. Was able to rebuild the library and perf_test with intel/compilers/17.2.174 without errors.

ndellingwood commented:
For reference, here's how I generate the makefile:
1. Load intel module for mkl 2017 or 2018
2. `export MKL_PATH=$MKLROOT`
3. Add this to a script and execute in some testing directory (the -mkl cppflag was necessary to get perf_tests compiling with proper mkl paths, thanks @kyungjoo-kim ! ):
```
KOKKOS_PATH=${HOME}/kokkos #path to kokkos source
KOKKOSKERNELS_SCALARS=double #the scalar types to instantiate =double,float...
KOKKOSKERNELS_LAYOUTS=LayoutLeft #the layout types to instantiate.
KOKKOSKERNELS_ORDINALS=int #ordinal types to instantiate
KOKKOSKERNELS_OFFSETS=int #offset types to instantiate
KOKKOSKERNELS_PATH=../.. #path to kokkos-kernels top directory.
CXX=icpc #${KOKKOS_PATH}/bin/nvcc_wrapper #icpc #
KOKKOSKERNELS_OPTIONS=eti-only #options for kokkoskernels
KOKKOS_DEVICES=OpenMP,Serial # other devices Cuda,Serial ..
KOKKOS_ARCHS=KNL
KOKKOSKERNELS_ENABLE_TPLS=mkl
CXXFLAGS="-Wall -pedantic -Werror -O3 -g -Wshadow -Wsign-compare -Wtype-limits -Wuninitialized -mkl"
```

4. `make install-lib -j16`
Note: Using -j with too many threads can cause problems with ranlib

5. `cd perf_tests; make -j`

ndellingwood commented:
Testbeds that I have access to with HSW + intel MKL versions for testing are pretty backed up right now for testing.

kyungjoo-kim commented:
The last commit will fix the warnings from perf_test. I confirmed that from white.

srajama1 commented:
I wish this additional commit was a separate one. This PR is addressing a number of issues now. I will review the latest one. @kyungjoo-kim : spot check on the usual suspects please.
